OpenRA.Mods.RA/Traits/Disguise.cs

```diff
@@ -177,6 +177,34 @@ void DisguiseAs(Actor target)
 			}
 		}
 
+		public void DisguiseAs(ActorInfo actorInfo, Player newOwner)
+		{
+			var oldDisguiseSetting = Disguised;
+			var oldEffectiveOwner = AsPlayer;
+
+			var renderSprites = actorInfo.Traits.GetOrDefault<RenderSpritesInfo>();
+			AsSprite = renderSprites == null ? null : renderSprites.GetImage(actorInfo, self.World.Map.SequenceProvider, newOwner.Country.Race);
+			AsPlayer = newOwner;
+			AsTooltipInfo = actorInfo.Traits.WithInterface<TooltipInfo>().FirstOrDefault();
+
+			foreach (var t in self.TraitsImplementing<INotifyEffectiveOwnerChanged>())
+				t.OnEffectiveOwnerChanged(self, oldEffectiveOwner, AsPlayer);
+
+			if (Disguised != oldDisguiseSetting && um.Value != null)
+			{
+				foreach (var u in info.Upgrades)
+				{
+					if (!um.Value.AcknowledgesUpgrade(self, u))
+						continue;
+
+					if (Disguised)
+						um.Value.GrantUpgrade(self, u, this);
+					else
+						um.Value.RevokeUpgrade(self, u, this);
+				}
+			}
```

Comment:
Can you factor the duplicated parts out into a separate private method, please? Thanks.